extensions-core/druid-kerberos/src/main/java/io/druid/security/kerberos/KerberosEscalator.java

```diff
@@ -59,4 +68,21 @@ public AuthenticationResult createEscalatedAuthenticationResult()
   {
     return new AuthenticationResult(internalClientPrincipal, authorizerName, null);
   }
+
+  @Override
+  public void decorateProxyRequest(
```

Comment:
This method should be part of the Authenticator instead of the Escalator.

Suppose a user has Kerberos authentication running alongside another authentication scheme, and they wish to use that other authentication scheme for Druid's internal communications. In that case, the Escalator would be of the other authentication type, and this Kerberos code would not be executed.

I think you could indicate the name of the Authenticator that checked a request (maybe put it in the AuthenticationResult), and in AsyncQueryForwardingServlet, get an injected copy of the AuthenticatorMapper and use the Authenticator name to lookup which Authenticator instance should be used to decorate the proxy request.

Reply:
@jon-wei Thanks good point was not aware of that. Will use that map that has the chain and call decorate on each one of them, that should work i guess.